compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -1719,43 +1808,174 @@ private BType checkArrayType(BLangListConstructorExpr listConstructor, BArrayTyp
+
+                        for (int i = remainNonRestCount; i < spreadOpMemberTypeSize; i++) {
+                            if (types.typeIncompatible(spreadOpExpr.pos, spreadOpTuple.tupleTypes.get(i), restType)) {
+                                return symTable.semanticError;
+                            }
+                        }
+                        continue;
+                    }
 
-        int nonRestCountToCheck = listExprSize < memberTypeSize ? listExprSize : memberTypeSize;
+                    if (spreadOpMemberTypeSize < remainNonRestCount) {
+                        dlog.error(spreadOpExpr.pos, DiagnosticErrorCode.INVALID_SPREAD_OP_FIXED_MEMBER_EXPECTED,
+                                memberTypes.get(nonRestTypeIndex + spreadOpMemberTypeSize));
+                        return symTable.semanticError;
+                    }
+
+                    for (int i = 0; nonRestTypeIndex < memberTypeSize; i++, nonRestTypeIndex++) {
+                        if (types.typeIncompatible(spreadOpExpr.pos, spreadOpTuple.tupleTypes.get(i),
+                                memberTypes.get(nonRestTypeIndex))) {
+                            return symTable.semanticError;
+                        }
+                    }
+
+                    for (int i = nonRestTypeIndex; i < spreadOpMemberTypeSize; i++) {
+                        if (types.typeIncompatible(spreadOpExpr.pos, spreadOpTuple.tupleTypes.get(i), restType)) {
+                            return symTable.semanticError;
+                        }
+                    }
 
-        for (int i = 0; i < nonRestCountToCheck; i++) {
-            if (exprIncompatible(memberTypes.get(i), exprs.get(i)) && !errored) {
-                errored = true;
+                    if (types.typeIncompatible(spreadOpExpr.pos, spreadOpTuple.restType, restType)) {
+                        return symTable.semanticError;
+                    }
+                    break;
+                default:
+                    dlog.error(spreadOpExpr.pos, DiagnosticErrorCode.INCOMPATIBLE_TYPES_LIST_SPREAD_OP, spreadOpType);
+                    return symTable.semanticError;
             }
         }
 
-        for (int i = nonRestCountToCheck; i < exprs.size(); i++) {
-            if (exprIncompatible(restType, exprs.get(i)) && !errored) {
-                errored = true;
+        while (nonRestTypeIndex < memberTypeSize) {
+            if (!types.hasFillerValue(memberTypes.get(nonRestTypeIndex))) {
+                dlog.error(listConstructor.pos, DiagnosticErrorCode.INVALID_LIST_CONSTRUCTOR_ELEMENT_TYPE,
+                        memberTypes.get(nonRestTypeIndex));
+                return symTable.semanticError;
             }
+            nonRestTypeIndex++;
         }
+
         return errored ? symTable.semanticError : tupleType;
     }
 
+    private boolean isFixedLengthTuple(BTupleType bTupleType) {
+        return bTupleType.restType == null || Types.getReferredType(bTupleType.restType).tag == TypeTags.NEVER;
```

Comment:
bTupleType.restType can ber record with never field; isn't it. shall we use `types.isNeverTypeOrStructureTypeWithARequiredNeverMember()` instead